server/protocols/jmap-integration-testing/memory-jmap-integration-testing/src/test/java/org/apache/james/jmap/memory/MemorySetMessagesMethodTest.java

```diff
@@ -48,4 +50,11 @@ protected void await() {
     protected MessageId randomMessageId() {
         return new InMemoryMessageId.Factory().fromString(String.valueOf(random.nextInt(100000) + 100));
     }
+
+    @Ignore("JAMES-2363 Bug: memory implementation should preserver mime message ids")
+    @Override
+    @Test
+    public void mimeMessageIdShouldBePreservedWhenSending() {
```

Comment:
TODO

Reply:
I found the bug after a debbugging session, and **it's tricky!**

The problem is that **RemoveHeader** mailet calls **MimeMessage::saveChanges()**, which in turn calls **updateHeaders**, which then generate a new **Message-Id**....

Reply:
An alternative approach (less invasive, but relies on inheritance) would be to rely on a custom implementation of **updateHeaders()** in **MimeMessageCopyOnWriteProxy**

Reply:
It is really depressing, these methods with unexpected side effects are called everywhere... Even a **setHeader** on MimeMessageCopyOnWriteProxy triggers this...

Reply:
I finally ended up replacing **MimeMessage::saveChanges()** calls by a **MimeMessageSaver::save()** which preserves message id.